### ISSUE TITLE
`button-has-type` not recognizing default props

### ISSUE THREAD

**Issue by pascalduez** [yannickcr/eslint-plugin-react#1846]
Hi,

looks like whatever I try the `button-has-type` rule can't understand the default prop.

```jsx
type Props = {
  type: 'submit' | 'button' | 'reset',
};

function Button({ type, ...extraProps }: Props) {
  return <button type={type} {...extraProps} />;
}

Button.defaultProps = {
  type: 'submit',
};
````

```
error  "null" is an invalid value for button type attribute  react/button-has-type
```

```jsx
type Props = {
  type: 'submit' | 'button' | 'reset',
};

function Button({ type = 'submit', ...extraProps }: Props) {
  return <button type={type} {...extraProps} />;
}
````

```
error  "null" is an invalid value for button type attribute  react/button-has-type
```

**Comment by ljharb** [comment]
How are you creating the `<Button>` element? Note that null doesn’t trigger defaults, but `undefined` does.

**Comment by ljharb** [comment]
Oh, i see that this is the linter reporting it. This seems like a bug.

**Comment by psa-gbellencontre** [comment]
I'm getting the same error. 
It seems that the button type attribute cannot be dynamic. But I don't understand why...

**Comment by ljharb** [comment]
It is required to be static for a number of reasons; one is that it’s lintable, but also because each kind of button is a fundemantally different component/element - they’re not just different “flavors” of button.

**Comment by truckingsim** [comment]
@ljharb In our component library at my company we have a Button component that is used for various things.  We use a `oneOf` prop type to do an enum of type options and default to button since that's the sane default.  But the button can be used technically as a submit or reset button (and does get used).  

We have a component due to css and classes, its easier to have one shared one where the app people don't have to worry about css classes and semantics, they can just use our Button.

I don't want to turn this rule off, but if its going to be forced to be static, don't really see another way.

**Comment by ljharb** [comment]
@truckingsim what i'd expect is that you'd have that Button component wrapped by SubmitButton and ResetButton components, and inside *those* you'd override the rule for the dynamic `type` - but you wouldn't need that anywhere else.

(ofc, this is separate from the issue that reset buttons are terrible UX and shouldn't ever be on the page)

**Comment by truckingsim** [comment]
@ljharb that's what I just did, but also feels like I shouldn't have to do that.

Is there no way to take props into consideration for rules?  So that basically if there's a default + it only allows what is expected it passes?

**Comment by ljharb** [comment]
If you can come up with a way for the rule to assert that the prop has a propType (or flow/TS type) that only permits one of the proper values, and that that prop is the dynamic value passed in, then that absolutely seems like something that could be allowed.

I'm not sure how to implement that, but I'd be happy to review a PR that achieves that.

**Comment by golopot** [comment]
Is it acceptable for the rule to just bail out when type is a curly expression?

**Comment by ljharb** [comment]
Not by default; the intention is for the rule to forced a hardcoded inline string type.

**Comment by JeromeLin** [comment]
![image](https://user-images.githubusercontent.com/9812721/58404130-fecfed00-8096-11e9-850c-caab2d9ba12a.png)

It's like taking the variable name as a string value.

**Comment by ljharb** [comment]
@JeromeLin https://github.com/yannickcr/eslint-plugin-react/issues/1846#issuecomment-401849676 ; we should make that error message better but it’d still need to be static.

**Comment by codedart2018** [comment]
eeeee
iam getting the same error. 好烦！

**Comment by peoplespete** [comment]
A coworker of mine just hit this same issue. She figured out two ways around the issue.

While all of the following feels hacky, these options might help until this bug is fixed.

Option 1: Store your button type in a variable named for one of the valid button types.
```
function Button({ type, ...extraProps }: Props) {
  const button = type;
  return <button type={button} {...extraProps} />;
}
```

Option 2: Disable this linter rule just for this line of code:
```
function Button({ type, ...extraProps }: Props) {
  // eslint-disable-next-line react/button-has-type
  return <button type={type} {...extraProps} >;
}
```

Happy coding!

**Comment by ljharb** [comment]
@peoplespete note that Option 1 working is actually a bug, which will be fixed very soon :-) Option 2 is the proper workaround.

**Comment by makoto-engineer** [comment]
Has it been fixed in this version?
```
"eslint-plugin-react": "7.27.1",
```

**Comment by ljharb** [comment]
@makoto-engineer yes, it was fixed in v7.20.0.

**Comment by ljharb** [comment]
I think this should be closed.

`<button type` values should always be static so they can be statically analyzed, and for UX reasons `reset` should never be used.

**Comment by Manubi** [comment]
Hi,
I think it's not fixed. I am using `eslint 8.14.0`
<img width="912" alt="CleanShot 2022-05-10 at 13 17 28@2x" src="https://user-images.githubusercontent.com/20935138/167617210-1ae0e939-548b-4a6d-912a-ef6111b6d627.png">
.

**Comment by ljharb** [comment]
@Manub it’s not fixable. Type values should be static, not coming from a prop.

**Comment by SalahAdDin** [comment]
> @manub it’s not fixable. Type values should be static, not coming from a prop (and “reset” should never ever be used)

Why?

Button html tag receive it like props, why should not we add it too?

**Comment by ljharb** [comment]
@SalahAdDin why to which part? "reset" is a horrible UX pattern that should never be used, if that's what you mean.

**Comment by SalahAdDin** [comment]
> @SalahAdDin why to which part? "reset" is a horrible UX pattern that should never be used, if that's what you mean.

I mean, the type property should be also available to be passed from a parameter.

**Comment by Manubi** [comment]
Please, don't hang me up on the reset button. I am talking about `type=button` and `type=submit`. That's all.. :)

**Comment by ljharb** [comment]
@Manubi @SalahAdDin it does _work_ just fine when dynamically provided. However, that makes it impossible to lint for it - to statically analyze it.

By hardcoding it, you allow the lint rule to be useful.

**Comment by Manubi** [comment]
but than I would need to have two button components one with type submit and one with type button. Right?

**Comment by ljharb** [comment]
@Manubi no, you could still dynamically render one of two button _elements_ within the same component.
